common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,915 @@
+        if (partitionInfo == null) {
+          if (isTooLate) {
+            // Throw an exception here so the block data is drained from channel and server
+            // responds RpcFailure to the client.
+            throw new RuntimeException(String.format("Block %s %s", msg.blockId,
+              ErrorHandler.BlockPushErrorHandler.TOO_LATE_MESSAGE_SUFFIX));
+          } else {
+            // For duplicate block that is received before the shuffle merge finalizes, the
+            // server should respond success to the client.
+            return;
+          }
+        }
+        // TODO should the merge manager check for the merge completion ratio here and finalize
+        // TODO shuffle merge if appropriate? So the merge manager can potentially finalize early
+        // TODO and the file channel can be closed even if finalize merge request is somehow not
+        // TODO received from the driver? If so, then we need to know # maps for this shuffle.
+
+        synchronized (partitionInfo) {
+          // When this request initially got to the server, the shuffle merge finalize request
+          // was not received yet. By the time we finish reading this message, the shuffle merge
+          // however is already finalized. We should thus respond RpcFailure to the client.
+          if (!partitions.containsKey(partitionId)) {
+            deferredBufs = null;
+            throw new RuntimeException(String.format("Block %s %s", msg.blockId,
+              ErrorHandler.BlockPushErrorHandler.TOO_LATE_MESSAGE_SUFFIX));
+          }
+          // Check if we can commit this block
+          if (allowedToWrite()) {
+            startBlockWrite = true;
+            // Identify duplicate block generated by speculative tasks. We respond success to
+            // the client in cases of duplicate even though no data is written.
+            if (isDuplicateBlock()) {
+              deferredBufs = null;
+              return;
+            }
+            if (partitionInfo.getCurrentMapId() < 0) {
+              writeAnyDeferredBlocks();
+            }
+            long updatedPos = partitionInfo.getPosition() + length;
+            boolean indexUpdated = false;
+            if (updatedPos - partitionInfo.getLastChunkOffset() >= minChunkSize) {
+              partitionInfo.updateChunkInfo(updatedPos, mapId);
+              indexUpdated = true;
+            }
+            partitionInfo.setPosition(updatedPos);
+            partitionInfo.setCurrentMapId(-1);
+
+            // update merged results
+            partitionInfo.blockMerged(mapId);
+            if (indexUpdated) {
+              partitionInfo.resetChunkTracker();
+            }
+          } else {
+            deferredBufs = null;
+            throw new RuntimeException(String.format("%s %s to merged shuffle",
+                ErrorHandler.BlockPushErrorHandler.BLOCK_APPEND_COLLISION_DETECTED_MSG_PREFIX,
+                msg.blockId));
```

Comment:
I guess this is the worst-case during the merging process. Do you have any statistics show how often it happens?

Another possible way to handle this case I'm thinking of is to write the deferred blocks into separate files and merge them into the merged data file when we receive the finalizing requests. Thus, we can avoid duplicate transfers. And since client sides also reading those blocks from files, so I think the only overhead we might introduce here is writing new files.

But I'm not sure which way could be better. Just brainstorm here.

Reply:
We don't have the exact stats. However, for a shuffle with larger number of partitions this happens often.
If we write the data into separate files, another overhead is that the time taken for finalization will increase. During finalization the shuffle server, would need to open all these other files, read the data, append it to the merged file, and then delete these other files. This would take more time.
The driver currently waits for a configured amount of time, before it moves on to the next stage. Right now the default for this is 10s. If the shuffle server is doing more work during finalization, then probably this timeout needs to be increased. So, eventually it ends up increasing the overall time of a job.
Since push-merge is best-effort, we opted for just discarding these `bufs`.
@Victsm

Reply:
> If we write the data into separate files, another overhead is that the time taken for finalization will increase. 

Make sense to me. Then, how about we merging them at the time the `onComplete()` is called?

Reply:
If a stream is `allowedToWrite`, it does write the `deferredBufs` to the merged file. It's line 551-553.
It's only when the stream is `completing` and still doesn't get a chance then it lands here. Even if we write these deferred blocks to separate files, they cannot be merged to the main data file because some other stream is in between writing. Merging can only happen during `onComplete` when the stream is `allowedToWrite`. If it is not, then it cannot merge.

Reply:
> If a stream is allowedToWrite, it does write the deferredBufs to the merged file during onComplete. It's line 551-553.

This's a different case comparing to the worst-case, right? If the `deferredBufs` still can't get a chance to merge, then, it results in the worst-case. And we're discussing how to merge those worst-case blocks if they have been written into separate files.

> Even if we write these deferred blocks to separate files, they cannot be merged to the main data file because some other stream is writing. Merging can only happen during onComplete when the stream is allowedToWrite.

So, there must be at least one thread win to write/merge the block into the data file, right? If so, we can always merge those separate files to the data file after the block is merged but before the thread releases the lock, right?

Reply:
Let's take an example with 2 blocks: block1 is `shuffle_1_1_1`  and block2 is `shuffle_1_2_1`. The shuffle service should try to append both of these  to `mergedShuffle_1_1.data`. 

When the shuffle service  starts receiving data for `block1` in a stream, let's say this is `stream1`, it appends the received `buf` which is part of `shuffle_1_1_1` to `mergedShuffle_1_1.data` and sets currentMapId = 1. Now before more data is received with `stream1`, the shuffle service receives data for `block2` with `stream2`. When `onData` of `stream2` is invoked, the `buf` will be deferred. Any subsequent `onData` on `stream2` will keep deferring the `bufs`.

The worst case happens when the `onComplete` of `stream2` is invoked before `onComplete` of `stream1`.
When this happens `stream2` cannot merge the deferred data  to `mergedShuffle_1_1.data` irrespective of where it saves the `deferredBufs`. Right now it's saving it in memory. However, even if it is saving the `deferredBufs` in file, it cannot append it to the `mergedShuffle_1_1.data` because `stream1` hasn't completed yet.

Reply:
> However, even if it saves the deferredBufs in file, it cannot append it to the mergedShuffle_1_1.data because stream1 hasn't completed yet.

We can merge the separate file at the time when `stream1` invokes `onComplete`(specifically, after `steam1` has been merged to the `mergedShuffle_1_1.data`) , can't we?

Reply:
So you are proposing a stream to access data of other streams.
In this example, `stream1` accesses the data of `stream2`. A natural question that follows this is how does `stream1` know that the data in `stream2` is complete?
There are ways to handle it and below are some that I can think of:
- Even if `stream2` writes deferred bufs to its own file  when it reaches this worst-case  during `onComplete`, we can't have `stream1` reading from this file while `stream2` is writing. So, `stream2` can write to tmp file and when it has completely written all the bufs it renames the tmp file. Since a rename operation is atomic, it will guarantee the consistency of `block2`.
- When `stream2` completes and reaches this worst-case, it marks it in the `appPartitionInfo`. Let's say the `appParitionInfo` contains a list of file location of completed streams which did not get a chance to merge. When `stream1` is done writing it checks this list for any other stream data it can merge. If yes, then it writes them and empties the list. This list should be thread-safe.

The main problem that I see with this approach where a stream is writing the data of other streams is that it increases the duration of the stream. In this example `stream1` will own the merged data file much longer. This may have a cascading impact. Other streams that have to write to same shuffle partition will have an increased chance of reachin this worst-cases.
Another issue is that there are still cases even with this approach, where the data of streams is ignored.  In this example, when `stream2` is writing its deferred blocks to a file while `stream1` is just about to complete. After `stream1` no other stream get to write.

Reply:
> The main problem that I see with this approach where a stream is writing the data of other streams is that it increases the duration of the stream. In this example stream1 will own the merged data file much longer. This may have a cascading impact. Other streams that have to write to same shuffle partition will have an increased chance of reachin this worst-cases.

I think this's fine. We can imagine or assume the `block2` is sending from the client aiagin and win the lock to write to the merged data file. Although, it's actually read from the separate file.

For example, previously, when the server returns a collision error of `block2` to the client, the `block2` will retry at some time later. And it's possible the retry `block2` wins the lock to write to the merged data file while deferring other blocks. In this case, you can just imagine the new proposal, `block2` doesn't go through the retry process but win the lock in the first place.

> Another issue is that there are still cases even with this approach, where the data of streams is ignored. In this example, when stream2 is writing its deferred blocks to a file while stream1 is just about to complete. After stream1 no other stream get to write.


hmm..I don't understand the issue here clearly...

The writing behavior of `stream1` and `stream2` is protected by the lock of the same parititonInfo. So, `stream2`  writing to the separate file and `stream1` is about to complete can not happen at the same time, right?

Reply:
> The writing behavior of stream1 and stream2 is protected by the lock of the same parititonInfo. So, stream2 writing to the separate file and stream1 is about to complete can not happen at the same time, right?

Yes, you are right. I missed that it was synchronizing on `partitionInfo`.

> I think this's fine. We can imagine or assume the block2 is sending from the client aiagin and win the lock to write to the merged data file. Although, it's actually read from the separate file.

I do think that if a stream starts writing deferred blocks of other streams to the merged block, it will increase the time that this stream will hold right to merge. This would mean that more streams would be reaching the worst case and writing data in separate files. This increases the I/O. The I/O will be significantly more since more streams are reaching this worst case now and fewer streams will be merging the data.

Reply:
> This would mean that more streams would be reaching the worst case and writing data in separate files.

You can imagine that it's not the `stream1` that increases the time to merge, but it's the time combines of `stream1` merging time and `stream2` merging time(`stream2` wins the lock (there're might be another `stream3` races with `stream2` for the lock but fails and is deferred to the file) right after the `stream1` finishes merging). 
 Things would not change even if `stream2` is re-sent from the client. Because `stream2` may also need to race with stream3 and wins the lock and so stream3 get deferred.

Reply:
> You can imagine that it's not the stream1 that increases the time to merge, but it's the time combines of stream1 merging time and stream2 merging time

Hmmm. There is additional IO now. stream1 will be reading the data of stream2 from disk which adds considerable time.

Also, the main reason behind push-based shuffle is to convert small random reads to larger sequential reads. The approach of writing individual blocks to disk and then later reading them again will result in a larger random reads.

Reply:
IIRC, the `block2` is also stored in the shuffle data file at the client-side? So, if we choose to re-send the `block2`, we will also read it again from the disk, right? So I think the trade-off here is read IO + network transfer (retry) vs write IO + read IO (write to the file). As I mentioned earlier, I really can not tell which way is obviously better without any experiment.

Reply:
We do want to reduce the small random reads by the shuffle service. The shuffle server is a shared resource serving shuffle data for multiple apps and these small random reads impacts its performance severely. Please refer to the  push-based shuffle SPIP [doc](https://docs.google.com/document/d/1mYzKVZllA5Flw8AtoX7JUcXBOnNIDADWRbJ7GI6Y71Q/edit) which mentions this as a goal. 
An executor reading the block again from the file would be much better than increasing the random reads on the shuffle server.